src/circuit/ecc/general_ecc.rs

```diff
@@ -4,107 +4,127 @@ use crate::circuit::main_gate::{MainGate, MainGateInstructions};
         &self,
         region: &mut Region<'_, N>,
         c: &AssignedCondition<N>,
-        p1: &AssignedPoint<Emulated, N>,
-        p2: &AssignedPoint<Emulated, N>,
+        p1: &AssignedPoint<N>,
+        p2: &AssignedPoint<N>,
         offset: &mut usize,
-    ) -> Result<AssignedPoint<Emulated, N>, Error>;
+    ) -> Result<AssignedPoint<N>, Error>;
 
     fn select_or_assign(
         &self,
         region: &mut Region<'_, N>,
         c: &AssignedCondition<N>,
-        p1: &AssignedPoint<Emulated, N>,
+        p1: &AssignedPoint<N>,
         p2: Emulated,
         offset: &mut usize,
-    ) -> Result<AssignedPoint<Emulated, N>, Error>;
+    ) -> Result<AssignedPoint<N>, Error>;
 
     fn assert_equal(
         &self,
         region: &mut Region<'_, N>,
-        p0: &AssignedPoint<Emulated, N>,
-        p1: &AssignedPoint<Emulated, N>,
+        p0: &AssignedPoint<N>,
+        p1: &AssignedPoint<N>,
         offset: &mut usize,
     ) -> Result<(), Error>;
 
     fn add(
         &self,
         region: &mut Region<'_, N>,
-        p0: &AssignedPoint<Emulated, N>,
-        p1: &AssignedPoint<Emulated, N>,
+        p0: &AssignedPoint<N>,
+        p1: &AssignedPoint<N>,
         offset: &mut usize,
-    ) -> Result<AssignedPoint<Emulated, N>, Error>;
+    ) -> Result<AssignedPoint<N>, Error>;
 
-    fn double(&self, region: &mut Region<'_, N>, p: AssignedPoint<Emulated, N>, offset: &mut usize) -> Result<AssignedPoint<Emulated, N>, Error>;
+    fn double(&self, region: &mut Region<'_, N>, p: AssignedPoint<N>, offset: &mut usize) -> Result<AssignedPoint<N>, Error>;
 
     fn mul_var(
         &self,
         region: &mut Region<'_, N>,
-        p: AssignedPoint<Emulated, N>,
+        p: AssignedPoint<N>,
         e: AssignedInteger<Emulated::ScalarExt>,
         offset: &mut usize,
-    ) -> Result<AssignedPoint<Emulated, N>, Error>;
+    ) -> Result<AssignedPoint<N>, Error>;
 
     fn mul_fix(
         &self,
         region: &mut Region<'_, N>,
         p: Point<N>,
         e: AssignedInteger<Emulated::ScalarExt>,
         offset: &mut usize,
-    ) -> Result<AssignedPoint<Emulated, N>, Error>;
+    ) -> Result<AssignedPoint<N>, Error>;
 }
 
 pub struct GeneralEccChip<Emulated: CurveAffine, F: FieldExt> {
-    config: EccConfig,
-    rns_base_field: Rns<Emulated::Base, F>,
-    rns_scalar_field: Rns<Emulated::Scalar, F>,
+    pub (super) config: EccConfig,
+    pub (super) rns_base_field: Rns<Emulated::Base, F>,
+    pub (super) rns_scalar_field: Rns<Emulated::Scalar, F>,
+
+    // We need following assigned constants for ecc operations
+    pub (super) a: AssignedInteger<F>,
+    pub (super) b: AssignedInteger<F>,
+    pub (super) identity: AssignedPoint<F>,
 }
 
+// Ecc operation mods
+mod add;
+
 impl<Emulated: CurveAffine, N: FieldExt> GeneralEccChip<Emulated, N> {
-    fn new(config: EccConfig, rns_base_field: Rns<Emulated::Base, N>, rns_scalar_field: Rns<Emulated::ScalarExt, N>) -> Self {
-        Self {
+    pub (super) fn new(
```

Comment:
While contructing the Ecc chip curve parameters `a`, `b` and `identity` assigned as witnesses for later use. Problem is that synthesized circuit will not define the circuit we would like to use. Then prover can assign any values they like. Instead of prover tiem assigned those parameters we need to assign those in synthesis time. So that these constants will take place in selector columns rather than witness columns. New improvements in master allows you to do so for example with add_constant functions.

Reply:
Adjusted.